File: swift/Sources/HermesAPI/Runtime/HermesAuth.swift
import Foundation

/// The credential to present on a single WebSocket upgrade.
public enum GatewayCredential: Sendable {
    case ticket(String, headers: [String: String])
    case localToken(String, headers: [String: String])
}

/// Supplies a fresh WebSocket credential for every connection or reconnect.
public protocol HermesAuth: Sendable {
    func credential(baseURL: URL, http: any HTTPTransport) async throws -> GatewayCredential
}

/// Authenticated dashboard ticket flow. A ticket is single-use and expires after 30 seconds.
public struct DashboardTicketAuth: HermesAuth {
    public let headers: @Sendable () async throws -> [String: String]

    public init(headers: @escaping @Sendable () async throws -> [String: String]) {
        self.headers = headers
    }

    public func credential(baseURL: URL, http: any HTTPTransport) async throws -> GatewayCredential {
        let headers = try await headers()
        guard let url = URL(string: "/api/auth/ws-ticket", relativeTo: baseURL)?.absoluteURL else {
            throw HermesGatewayError.transport("Invalid dashboard URL")
        }
        var request = URLRequest(url: url)
        request.httpMethod = "POST"
        for (name, value) in headers { request.setValue(value, forHTTPHeaderField: name) }
        let (data, response) = try await http.send(request)
        guard (200..<300).contains(response.statusCode) else {
            throw HermesGatewayError.transport("WebSocket ticket request returned HTTP \(response.statusCode)")
        }
        let ticket = try JSONDecoder().decode(TicketResponse.self, from: data).ticket
        guard !ticket.isEmpty else { throw HermesGatewayError.decoding("Empty WebSocket ticket") }
        return .ticket(ticket, headers: headers)
    }
}

private struct TicketResponse: Decodable {
    let ticket: String
}

/// The local dashboard's legacy session token. Gated dashboards reject this credential.
public struct LocalTokenAuth: HermesAuth {
    public let token: String
    public let headers: [String: String]

    public init(token: String, headers: [String: String] = [:]) {
        self.token = token
        self.headers = headers
    }

    public func credential(baseURL: URL, http: any HTTPTransport) async throws -> GatewayCredential {
        .localToken(token, headers: headers)
    }
}

File: kotlin/src/main/kotlin/st/flrn/hermes/api/runtime/HermesAuth.kt
package st.flrn.hermes.api.runtime

import java.net.URI
import kotlinx.serialization.json.Json
import kotlinx.serialization.json.jsonObject
import kotlinx.serialization.json.jsonPrimitive

public sealed interface GatewayCredential {
    public data class Ticket(val value: String, val headers: Map<String, String>) : GatewayCredential
    public data class LocalToken(val value: String, val headers: Map<String, String>) : GatewayCredential
}

/** A new credential is requested on each connection, including reconnects. */
public interface HermesAuth {
    public suspend fun credential(baseURI: URI, http: GatewayHTTPTransport): GatewayCredential
}

public class DashboardTicketAuth(private val headers: suspend () -> Map<String, String>) : HermesAuth {
    override suspend fun credential(baseURI: URI, http: GatewayHTTPTransport): GatewayCredential {
        val currentHeaders = headers()
        val (status, body) = http.post(baseURI.resolve("/api/auth/ws-ticket"), currentHeaders)
        if (status == 401 || status == 403) {
            throw HermesGatewayException.AuthenticationFailed("WebSocket ticket request returned HTTP $status")
        }
        if (status !in 200..299) throw HermesGatewayException.Transport("WebSocket ticket request returned HTTP $status")
        val ticket = Json.parseToJsonElement(body).jsonObject["ticket"]?.jsonPrimitive?.content
            ?: throw HermesGatewayException.Protocol("Ticket response has no ticket")
        if (ticket.isEmpty()) throw HermesGatewayException.Protocol("Empty WebSocket ticket")
        return GatewayCredential.Ticket(ticket, currentHeaders)
    }
}

public class LocalTokenAuth(private val token: String, private val headers: Map<String, String> = emptyMap()) : HermesAuth {
    override suspend fun credential(baseURI: URI, http: GatewayHTTPTransport): GatewayCredential =
        GatewayCredential.LocalToken(token, headers)
}
